crates/subspace-solving/src/codec.rs

```diff
@@ -133,16 +134,30 @@ impl SubspaceCodec {
 
         #[cfg(feature = "cuda")]
         if self.cuda_available {
-            let pieces_to_process = pieces.len() / PIECE_SIZE / GPU_PIECE_BLOCK * GPU_PIECE_BLOCK;
-
-            self.batch_encode_cuda(
-                &mut pieces[..pieces_to_process * PIECE_SIZE],
-                &piece_indexes[..pieces_to_process],
-            )?;
-
-            // Leave the rest for CPU
-            pieces = &mut pieces[pieces_to_process * PIECE_SIZE..];
-            piece_indexes = &piece_indexes[pieces_to_process..];
+            let mut pieces_to_process = pieces.len() / PIECE_SIZE;
+
+            // GPU will accept multiples of 1024 pieces
+            if pieces_to_process >= 1024 {
+                pieces_to_process = pieces_to_process / GPU_PIECE_BLOCK * GPU_PIECE_BLOCK;
+                // process the multiples of 1024 pieces in GPU
+                let cuda_result = self.batch_encode_cuda(
+                    &mut pieces[..pieces_to_process * PIECE_SIZE],
+                    &piece_indexes[..pieces_to_process],
+                );
+
+                if let Err(e) = cuda_result {
+                    error!("An error happened on the GPU: '{}'", e);
+                    self.cuda_available = false;
+                    // TODO: maybe also return from the GPU last successful encoding,
+                    // so that CPU can continue from there
+                    // because this implementation does not cover the case when GPU creates a problem
+                    // after some pieces are successfully encoded
```

Comment:
But GPU doesn't modify inputs unless encoding was done successfully, isn't it? I mean it copies data to GPU, encodes it and then copies back after successful encoding, right?

Reply:
No. Since we will be dealing with big chunks of data, GPU encodes each chunk, overwrites that chunk back, and proceeds to encode the next chunk.

Reply:
Then we need to have a way to distinguish if input was changed or not.

Reply:
That is why I suggested the TODO in the code. 
One way would be cloning the original and then comparing it with the GPU result. However, very inefficient...
Second one can be, after each successful iteration on the GPU side, GPU can update the index of where it is, and return that with the error message. 
If there is no error message, it will mean all the pieces are encoded. If there is an error message, we will know till where it was successful. And CPU can continue from there